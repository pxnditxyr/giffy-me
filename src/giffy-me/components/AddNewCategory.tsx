import { FormEvent, useState } from 'react';
import { useForm } from '../hooks/useForm';

interface AddNewCategoryProps {
    currentCategories: Array<string>;
    onAddNewCategory: ( category : string ) => void;
}

export const AddNewCategory = ( { currentCategories, onAddNewCategory } : AddNewCategoryProps ) => {

    const [ isSame, setIsSame ] = useState<Boolean>( false );
    
    const { category, handleInputChange, handleReset } = useForm({
        category: ''
    });

    const handleSubmit = ( event : FormEvent<HTMLFormElement> ) => {
        event.preventDefault();
        
        const newCategory = category.trim();

        if ( newCategory.length < 2 ) return;

        if ( currentCategories.map( currentCategory => currentCategory.toLowerCase() ).includes( newCategory.toLowerCase() ) ) {
            setIsSame( true );
            return;
        }

        if ( isSame ) setIsSame( false );

        onAddNewCategory( newCategory );
        handleReset();
    };

    return (
        <form
            onSubmit={ handleSubmit }
            >
            <label> What are you thinking </label>
            <input
                type="text"
                name="category"
                value={ category }
                onChange={ handleInputChange }
                />
            {
                isSame 
                    && 
                ( <p> You already have this category </p> )
            }
            <button type="button"> Search </button>
        </form>
    );
};
