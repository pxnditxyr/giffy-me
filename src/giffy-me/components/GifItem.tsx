
interface GifItemProps {
    // id: string;
    title: string;
    url: string;
}

export const GifItem = ( { title, url } : GifItemProps ) => {

    return (
        <article
            className="flex flex-col justify-center items-center"
        >
            <img
                src={ url }
                alt={ title }
                className="w-full h-auto object-cover rounded-lg"
                />
            <p
                className="text-center text-xl font-bold text-gray-300 py-4"
            > { title } </p>
        </article>
    );
};
