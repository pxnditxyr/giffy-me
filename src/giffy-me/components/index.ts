export * from './AddNewCategory';
export * from './GifGrid';
export * from './GifItem';
