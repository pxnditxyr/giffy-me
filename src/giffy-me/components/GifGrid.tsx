import { GifItem } from './GifItem';

import { useFetchGifs } from '../hooks/useFetchGifs';

interface GifGridProps {
    category: string;
}

export const GifGrid = ( { category } : GifGridProps ) => {

    const { gifs, isLoading } = useFetchGifs( category );

    return (
        <div
            className="w-full flex flex-col justify-center items-center px-5"
        >
            <h3 
                className="text-center text-4xl font-bold text-amber-400 py-4 md:py-20"
            > { category } </h3>
            {
                isLoading
                    ? ( <p> Loading... </p> )
                    : (
                        <div
                            className="grid gap-4 grid-cols-1 md:grid-cols-2 lg:grid-cols-3 px-4"
                        >
                            {
                                gifs.map( ( gif ) => (
                                    <GifItem 
                                        key={ gif.id }
                                        { ...gif }
                                    />
                                ))
                            }
                        </div>
                    )
            }
        </div>
    );
};
