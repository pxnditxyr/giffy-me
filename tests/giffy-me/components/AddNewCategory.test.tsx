import { fireEvent, render, screen } from '@testing-library/react';

import { AddNewCategory } from '../../../src/giffy-me/components/AddNewCategory';

describe( 'tests in AddNewCategory Component', () => {
    const currentCategories : Array<string> = [];

    test( 'should render AddNewCategory component', () => {

        const onAddNewCategory = jest.fn();
        const { container } = render( <AddNewCategory currentCategories={ currentCategories } onAddNewCategory={ onAddNewCategory } /> );
        expect( container ).toMatchSnapshot();
    });

    test( 'should execute onAddNewCategory function', () => {
        const onAddNewCategory = jest.fn();
        render( <AddNewCategory currentCategories={ currentCategories } onAddNewCategory={ onAddNewCategory } /> );
        const input : HTMLInputElement = screen.getByRole( 'textbox' );
        fireEvent.input( input, { target: { value: 'Naruto' } } );
        expect( input.value ).toBe( 'Naruto' );
    });

    test( 'should call onAddNewCategory if input has a value that has length greater than or equal to 2', () => {

        const onAddNewCategory = jest.fn();
        const inputValue = 'Naruto';

        render( <AddNewCategory currentCategories={ currentCategories } onAddNewCategory={ onAddNewCategory } /> );

        const input : HTMLInputElement = screen.getByRole( 'textbox' );
        const form : HTMLFormElement = screen.getByRole( 'form' );

        fireEvent.input( input, { target: { value: inputValue } } );
        fireEvent.submit( form );

        expect( input.value ).toBe( '' );

        expect( onAddNewCategory ).toHaveBeenCalled();
        expect( onAddNewCategory ).toHaveBeenCalledTimes( 1 );
        expect( onAddNewCategory ).toHaveBeenCalledWith( inputValue );
    });

    test( 'should not call onAddNewCategory if input has a value that has length less than 2', () => {

        const onAddNewCategory = jest.fn();
        const inputValue = 'x';
        
        render( <AddNewCategory currentCategories={ currentCategories } onAddNewCategory={ onAddNewCategory } /> )

        const input : HTMLInputElement = screen.getByRole( 'textbox' );
        const form : HTMLFormElement = screen.getByRole( 'form' );

        fireEvent.input( input, { target: { value: inputValue } } );
        fireEvent.submit( form );

        expect( onAddNewCategory ).not.toHaveBeenCalled();
    });

});
