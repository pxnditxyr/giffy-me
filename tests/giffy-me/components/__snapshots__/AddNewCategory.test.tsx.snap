// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests in AddNewCategory Component should render AddNewCategory component 1`] = `
<div>
  <form
    aria-label="form"
    class="flex flex-col items-center justify-center w-full py-3 px-8"
  >
    <div
      class="flex flex-col gap-8"
    >
      <label
        class="text-amber-400 font-semibold text-3xl"
      >
         What are you thinking? 
      </label>
      <div
        class="flex flex-col gap-5 md:flex-row"
      >
        <input
          class="bg-gray-800 text-white font-semibold text-3xl py-2 px-4 rounded-lg w-full focus:outline-none md:w-2/3"
          name="category"
          placeholder="Add a new category"
          type="text"
          value=""
        />
        <button
          class="bg-gray-800 text-white font-semibold text-3xl py-2 px-4 rounded-lg w-full md:w-1/3"
          type="submit"
        >
           Search 
        </button>
      </div>
    </div>
  </form>
</div>
`;
